### PR TITLE
Optimize EngineFileUtilities.GetFileList

### LINKED ISSUE #6061: Microsoft.Build.Internal.EngineFileUtilities.GetFileListEscaped is 3x slower for netcore projects
Child of #6034. Traces at "\\SLNATSCALE002\CoreVsFramework\Traces\Core_400.etl.zip"

For 400 projects it takes 960ms(netcore) vs 320ms(netframework) and the delta increases with project.

### PR DESCRIPTION
**Pull request by ladipro**
Fixes #6061 

### Context

`EngineFileUtilities.GetFileList` has been identified as one of the performance bottlenecks when building .NET Core projects. Since this code is called as part of project evaluation, it directly impacts Visual Studio performance as well, especially solution load.

### Changes Made

Tweaked the code under `GetFileList`:
- Smarter order of wildcard checks to optimize for the common case.
- Optimized hex number parsing and wildcard detection.

### Testing

Existing unit tests for correctness, ETL traces for performance.

This change together with #6151 is showing about 30% less time spent in this particular function when building a Hello World .NET Core project with the Framework version of MSBuild. It's an OK win for project evaluation perf but translates to less than 1 ms of total command line build time.

### Notes

This PR is small but still recommended to be reviewed commit by commit.

### CODE REVIEW

File: src/Build/Utilities/EngineFileUtilities.cs [hunk 1]
```diff
@@ -89,31 +87,21 @@ internal string[] GetFileListEscaped
...
-            {
-                return true;
-            }
-            else
+            if (!FileMatcher.HasWildcards(filespecEscaped))
             {
                 return false;
             }
+
+            // If the item's Include has both escaped wildcards and  real wildcards, then it's
```
Comment (Forgind) [inline src/Build/Utilities/EngineFileUtilities.cs:100]: You took out the color from the original! But that's ok 😉

**Review by Forgind — APPROVED** [review]
LGTM, thanks!

File: src/Shared/EscapingUtilities.cs [hunk 2]
```diff
@@ -85,18 +99,16 @@ internal static string UnescapeAll(string escapedString, bool trim = false)
...
                 {
                     // First copy all the characters up to the current percent sign into
                     // the destination.
                     unescapedString.Append(escapedString, currentPosition, indexOfPercent - currentPosition);
 
                     // Convert the %XX to an actual real character.
-                    string hexString = escapedString.Substring(indexOfPercent + 1, 2);
-                    char unescapedCharacter = (char)int.Parse(hexString, System.Globalization.NumberStyles.HexNumber,
-                        CultureInfo.InvariantCulture);
+                    char unescapedCharacter = (char)((digit1 << 4) + digit2);
```
Comment (BenVillalobos) [inline src/Shared/EscapingUtilities.cs:111]: nice!

File: src/Shared/EscapingUtilities.cs [hunk 3]
```diff
@@ -226,28 +226,26 @@ string unescapedString
...
-
-                            (-1 != escapedString.IndexOf("%2a", StringComparison.Ordinal)) ||
-                            (-1 != escapedString.IndexOf("%2A", StringComparison.Ordinal)) ||
-                            (-1 != escapedString.IndexOf("%3f", StringComparison.Ordinal)) ||
-                            (-1 != escapedString.IndexOf("%3F", StringComparison.Ordinal))
-                        ;
+                    // %3f or %3F
+                    return true;
                 }
+                index = escapedString.IndexOf('%', index + 1, escapedString.Length - index - 3);
```
Comment (BenVillalobos) [inline src/Shared/EscapingUtilities.cs:248]: Why subtract 3 here?

**Review by BenVillalobos — COMMENTED** [review]
Mostly LGTM

**Review by cdmihai — APPROVED** [review]

File: src/Shared/EscapingUtilities.cs (hunk 3 above)
Comment (Forgind) [reply src/Shared/EscapingUtilities.cs:248]: We want something that looks like %3f or whatever, and that takes three characters. This means no need for verifying that there are enough characters afterwards.

Comment (BenVillalobos) [reply src/Shared/EscapingUtilities.cs:248]: Ah,  index + 1 means 1 less character to look at ;)

**Review by BenVillalobos — APPROVED** [review]

File: src/Shared/EscapingUtilities.cs (hunk 3 above)
Comment (Forgind) [reply src/Shared/EscapingUtilities.cs:248]: Well, we also know that index is the index of a %, so using index instead would result in saying `index = index;` ad infinitum.

File: src/Shared/EscapingUtilities.cs [hunk 4]
```diff
@@ -214,28 +226,26 @@ string unescapedString
...
+            if (escapedString.Length < 3)
             {
-                // It has a '%' sign.  We have promise.
-                if (
-                        (-1 != escapedString.IndexOf("%2", StringComparison.Ordinal)) ||
-                        (-1 != escapedString.IndexOf("%3", StringComparison.Ordinal))
-                    )
+                return false;
+            }
+            int index = escapedString.IndexOf('%', 0, escapedString.Length - 2);
```
Comment (KirillOsenkov) [inline src/Shared/EscapingUtilities.cs:235]: Do we have good test coverage here?

Comment (Forgind) [reply src/Shared/EscapingUtilities.cs:235]: We certainly have [some](https://github.com/dotnet/msbuild/blob/0def9d13d74d075e62851630fe296bb019e401a1/src/Shared/UnitTests/EscapingUtilities_Tests.cs#L74-L85), though it wouldn't be bad to add a couple new cases.

Comment (ladipro) [reply src/Shared/EscapingUtilities.cs:235]: I have added a few more cases, thank you!

File: src/Shared/EscapingUtilities.cs (hunk 3 above)
Comment (ladipro) [reply src/Shared/EscapingUtilities.cs:248]: I tweaked the expression to be more readable and added comments.

**Pull request merged**